### ISSUE TITLE
Please specify the behavior of min > max in Clip

### ISSUE THREAD

**Issue by mgehre-amd** [onnx/onnx#6165]
# Ask a Question

### Question
[onnx.Clip](https://onnx.ai/onnx/operators/onnx__Clip.html) leaves it open what happens when values for min and max are specified such that min > max.
Should this
- be an error (invalid model)
- set all values to min
- set all values to max
- be unspecified 
 
This is relevant to implement compilers.

**Comment by justinchuby** [comment]
cc @gramalingam

**Comment by AlexandreEichenberger** [comment]
Scavenging the specs for other such "out of bounds" type situations.

### We often use the term `undefined` for ops that have bad inputs.

**Cast**
```
Converting non-numerical-literal string such as "Hello World!" is an undefined behavior. Cases of converting string representing floating-point arithmetic value, such as "2.718", to INT is an undefined behavior.
```

**Many ops using `axes` list**

```
Behavior is undefined if an axis is repeated. 
```

**Reduction ops**

```
Reduction over an empty set of values yields minus infinity (if supported by the datatype) or undefined otherwise.
```

### In other places, we use the term `error`

**Gather op family**
```
It is an error if any of the index values are out of bounds.
```

**Loop**

```
It is an error if the dimensions or data type of these scan_outputs change across loop iterations.
```

**GetElement**

```
 It is an error if the input is an empty optional-type (i.e. does not have an element) and the behavior is undefined in this case.
```

**Resize**
```
One of 'scales' and 'sizes' MUST be specified and it is an error if both are specified.
```

**Comment by AlexandreEichenberger** [comment]
In summary, my recommendation would be to 

1. Declare the case where `min > max` undefined, in conformity with other such case already listed in the specs (see above)
2. Have all the cases that "does not make sense" normalized to `undefined` (aka change `error` to `undefined`). And then define early in the specs a (non-exhaustive) list of possible actions when encountering these cases, such as "aborting a compilation or execution with an error message", "generating a warning and taking a (tool specific) defined action", "silently ignoring it and taking a (tool specific) defined action.

**Comment by gramalingam** [comment]
> In summary, my recommendation would be to
> 
> 1. Declare the case where `min > max` undefined, in conformity with other such case already listed in the specs (see above)
> 2. Have all the cases that "does not make sense" normalized to `undefined` (aka change `error` to `undefined`). And then define early in the specs a (non-exhaustive) list of possible actions when encountering these cases, such as "aborting a compilation or execution with an error message", "generating a warning and taking a (tool specific) defined action", "silently ignoring it and taking a (tool specific) defined action.

Sounds good. Wrt the last point: it would be good to hear from backend implementors what actions they would like to do in such cases. Specifically: when does it make sense to produce a legal output value and continue execution? Eg., in the clip case, ending execution with an error would be the safest option. But, I can imagine an argument for allowing execution to proceed (namely that AI/ML models are meant to deal with noise anyway, and maybe it is harder to do this in some SIMD-parallel-execution settings, etc.)

**Comment by mgehre-amd** [comment]
> > In summary, my recommendation would be to
> > 
> > 1. Declare the case where `min > max` undefined, in conformity with other such case already listed in the specs (see above)
> > 2. Have all the cases that "does not make sense" normalized to `undefined` (aka change `error` to `undefined`). And then define early in the specs a (non-exhaustive) list of possible actions when encountering these cases, such as "aborting a compilation or execution with an error message", "generating a warning and taking a (tool specific) defined action", "silently ignoring it and taking a (tool specific) defined action.
> 
> Sounds good. Wrt the last point: it would be good to hear from backend implementors what actions they would like to do in such cases. Specifically: when does it make sense to produce a legal output value and continue execution? Eg., in the clip case, ending execution with an error would be the safest option. But, I can imagine an argument for allowing execution to proceed (namely that AI/ML models are meant to deal with noise anyway, and maybe it is harder to do this in some SIMD-parallel-execution settings, etc.)

Especially in when offloading to accelerators, error propagation is cumbersome and can be slow. On the other hand, having a lot of undefined behavior makes it really hard to differentiate backend bugs from invalid models.
One option that seems to work well is to make the resulting values unspecified (but continue) to keep accelerator implementations simple and fast, and then also provide a way to run the model with explicit error checking (e.g. through the onnx reference implementation / onnxruntime / ...) - like C++ sanitizers.

**Comment by AlexandreEichenberger** [comment]
Right now, we seem to use `undefined` and `error` indiscriminately.  My suggestion is that we use in the spec `undefined` to indicate cases that are not supported by the SPECS. And to urge the backend vendors to define what they do when encountering `undefined` so that the users ultimately know what happens when a model with some operation with undefined behavior is encountered.

As you rightly pointed out in your responses, there might be multiple "valid" responses to encountering an operation with `undefined` behavior. In some cases, a best effort (aka ignoring) may be desired. In other case (e.g. on a validation device), a precise error would be valuable to the user.  

Ultimately, the goal is to better communicate to the user what is done when encountering models with undefined behavior. 

I struggle a bit with the name `undefined` but I think it makes sense. The behavior is `undefined` from the SPEC perspective in that there is not one expected correct outcome. And at the same time we can urge the backend vendors to precisely specify what the do when encountering `undefined models` so as to better "define" its behavior :-)

**Comment by AlexandreEichenberger** [comment]
As I started doing a PR to fix the documentation for clip, I realized that Numpy/Pytorch have a defined value for this case, whereas TensorFlow does not.

### Frameworks

Numpy and PyTorch
```
An array with the elements of a, but where values < a_min are replaced with a_min, and those > a_max with a_max.
```
https://numpy.org/doc/stable/reference/generated/numpy.clip.html
https://pytorch.org/docs/stable/generated/torch.clamp.html#torch.clamp

TensorFlow:
```
Note: clip_value_min needs to be smaller or equal to clip_value_max for correct results.
```
https://www.tensorflow.org/api_docs/python/tf/clip_by_value

### Implementation cost (when min/max are runtime values)

The "undefined" approach can use 2 compare operations in parallel and 2 sequential select operations. Implementing the PyTorch/Numpy approach can be implemented by a serialization (compare->select->compare->select) or by 1 additional compare (min/max) and one logical (or) of the compare result.

So should we revisit this issue in light of this?

**Comment by mgehre-amd** [comment]
Note that in the C++ world, `undefined` has a strong meaning where the whole model output can change upon encountering undefined behavior, *even* if the result of the op with undefined behavior isn't used. (i.e. the compiler can assume that undefined behavior will never occur).
A more benign term is `unspecified` behavior, where the spec doesn't say what will happen, but compilers are not allowed to assume that this behavior never occurs in models. And then you got `implementation-defined`, which is like `unspecified` behavior but where conforming implementations must says what their behavior is.

**Comment by AlexandreEichenberger** [comment]
@gramalingam Any inputs on what to do with min > max? As I listed below, the result is defined for numpy & PyTorch, unspecified for TensorFlow. There is a small change in implementation, depending on how it is done, the overhead would be quite minimal (see above for options).

**Comment by gramalingam** [comment]
Thanks for the investigation @AlexandreEichenberger. Given this, it seems to me that PyTorch/Numpy definition seems best (since it covers TF's undefined behavior also). However, the one question that remains is whether we should make this a clarification now ... or, say it is undefined in opset 13, and make it defined in next opset with the pytorch/numpy semantics. The latter is slightly better (in a procedural sense).

**Comment by AlexandreEichenberger** [comment]
> [...] it seems to me that PyTorch/Numpy definition seems best (since it covers TF's undefined behavior also)

good way to phrase it, makes sense.

> [...] the one question that remains is whether we should make this a clarification now.

I will defer to your judgement @gramalingam .

**Comment by AlexandreEichenberger** [comment]
@postrational Any preference Michal?

**Comment by gramalingam** [comment]
I suggest we just make it a clarification

**Comment by AlexandreEichenberger** [comment]
implemented in #6385

**Comment by mgehre-amd** [comment]
> implemented in #6385
@AlexandreEichenberger, is this the correct link?

**Comment by AlexandreEichenberger** [comment]
Migrated to #6395

**Comment by AlexandreEichenberger** [comment]
Valuable input here on "undefined" vs "unspecified", and probably we could tighten the specs using that insight.

Since the PR addressing the specific issue at hand (clip) is now merged, closing this issue.
